pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,363 @@
+		nsdObjects.ObjectsByNamespace = append(nsdObjects.ObjectsByNamespace, v1alpha1.NamespaceAndNames{
+			Namespace: key.NamespacedName.Namespace,
+			Names:     []string{key.NamespacedName.Name},
+		})
+
+		// update index mapping
+		nsObjectsLocationInSlice[gvrAndNSKey] = len(nsdObjects.ObjectsByNamespace) - 1
+		return
+	}
+
+	// GVR mapping isn't found, create new entry for this obj (and NS map)
+	// add namespaceScopeDownsyncObjectsMap to the workload
+	workload.NamespaceScope = append(workload.NamespaceScope, v1alpha1.NamespaceScopeDownsyncObjects{
+		GroupVersionResource: metav1.GroupVersionResource{
+			Group:    gvr.Group,
+			Version:  gvr.Version,
+			Resource: gvr.Resource,
+		},
+		ObjectsByNamespace: []v1alpha1.NamespaceAndNames{
+			{
+				Namespace: key.NamespacedName.Namespace,
+				Names:     []string{key.NamespacedName.Name},
+			},
+		},
+	})
+
+	// retain a pointer to the added namespaceScopeDownsyncObjectsMap for efficiency
+	namespaceScopeDownsyncObjectsMap[gvr] = &workload.NamespaceScope[len(workload.NamespaceScope)-1]
+	// update ns mapping
+	nsObjectsLocationInSlice[util.KeyFromGVRandNS(gvr, key.NamespacedName.Namespace)] = 0 // first entry
+}
+
+func (resolution *placementResolution) matchesPlacementDecisionSpec(placementDecisionSpec *v1alpha1.PlacementDecisionSpec,
+	gvkGvrMapper util.GvkGvrMapper) bool {
+	resolution.RLock()
+	defer resolution.RUnlock()
+
+	// check destinations
+	if !destinationsMatch(resolution.destinations, placementDecisionSpec.Destinations) {
+		return false
+	}
+
+	// check workload
+	return workloadMatchesPlacementDecisionSpec(&placementDecisionSpec.Workload, resolution.objectIdentifierToKeyMap, gvkGvrMapper)
+}
+
+// destinationsMatch returns true if the destinations in the resolution
+// match the destinations in the placement decision spec.
+// The function performs one pass over both slices, sequentially.
+func destinationsMatch(destinations []string, placementDecisionDestinations []v1alpha1.Destination) bool {
+	if len(destinations) != len(placementDecisionDestinations) {
+		return false
+	}
+
+	// calculate match by counting matching objects then comparing to length.
+	// this works since entries are unique.
+	matches := map[string]interface{}{}
+	matchesCounter := 0
+
+	// fill up matches
+	for _, destination := range placementDecisionDestinations {
+		matches[destination.ClusterId] = nil
+	}
+
+	// check matches
+	for _, destination := range destinations {
+		if _, found := matches[destination]; !found {
+			return false
+		}
+
+		matchesCounter++
+	}
+
+	return len(placementDecisionDestinations) == matchesCounter
+}
+
+// workloadMatchesPlacementDecisionSpec returns true if the workload in the
+// resolution matches the workload in the placement decision spec.
+func workloadMatchesPlacementDecisionSpec(placementDecisionSpecWorkload *v1alpha1.DownsyncObjectReferences,
+	objectIdentifierToKeyMap map[string]*util.Key, gvkGvrMapper util.GvkGvrMapper) bool {
+	// check lengths
```

Comment:
If PlacementDecision used a flat structure as in #1222 then this length calculation would be just one use of `len()`.

Reply:
The structure was inherited from SyncerConfig (@nirrozenbaum).

Reply:
Yes, I know.
This is only an FYI, not a blocker.